flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -0,0 +1,237 @@
+      val list = this.getRuntimeContext.
+        getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
+
+      weightVector = list.get(0)
+    }
+
+    override def map(example: LabeledVector): (WeightVector, Int) = {
+
+      val lossFunction = parameterMap(LossFunction)
+      //TODO(tvas): Should throw an error if Dimensions has not been defined
+      val dimensions = example.vector.size
+      // TODO(tvas): Any point in carrying the weightGradient vector for in-place replacement?
+      // The idea in spark is to avoid object creation, but here we have to do it anyway
+      val weightGradient = new DenseVector(new Array[Double](dimensions))
+
+      val (loss, lossDeriv) = lossFunction.lossAndGradient(example, weightVector, weightGradient)
+
+      // Restrict the value of the loss derivative to avoid numerical instabilities
+      val restrictedLossDeriv: Double = {
+        if (lossDeriv < -IterativeSolver.MAX_DLOSS) {
+          -IterativeSolver.MAX_DLOSS
+        }
+        else if (lossDeriv > IterativeSolver.MAX_DLOSS) {
+          IterativeSolver.MAX_DLOSS
+        }
+        else {
+          lossDeriv
+        }
+      }
+
+      (new WeightVector(weightGradient, restrictedLossDeriv), 1)
+    }
+  }
+
+  /** Performs the update of the weights, according to the given gradients and regularization type.
+    *
+    */
+  private class WeightsUpdate() extends
+  RichMapFunction[(WeightVector, Int), WeightVector] {
+
+
+    var weightVector: WeightVector = null
+
+    @throws(classOf[Exception])
+    override def open(configuration: Configuration): Unit = {
+      val list = this.getRuntimeContext.
+        getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
+
+      weightVector = list.get(0)
+    }
+
+    override def map(gradientsAndCount: (WeightVector, Int)): WeightVector = {
+      val regularizationType = parameterMap(RegularizationType)
+      val regularizationParameter = parameterMap(RegularizationParameter)
+      val stepsize = parameterMap(Stepsize)
+      val weightGradients = gradientsAndCount._1
+      val count = gradientsAndCount._2
+
+      // Scale the gradients according to batch size
+      BLAS.scal(1.0/count, weightGradients.weights)
+
+      val weight0Gradient = weightGradients.intercept / count
+
+      val iteration = getIterationRuntimeContext.getSuperstepNumber
+
+      // Scale initial stepsize by the inverse square root of the iteration number
+      // TODO(tvas): There are more ways to determine the stepsize, possible low-effort extensions
+      // here
+      val effectiveStepsize = stepsize/math.sqrt(iteration)
+
+      val newWeights = weightVector.weights.copy
+      // Take the gradient step
+      BLAS.axpy(-effectiveStepsize, weightGradients.weights, newWeights)
+      val newWeight0 = weightVector.intercept - effectiveStepsize * weight0Gradient
+
+      // Apply the regularization
+      val (updatedWeights, regVal) = regularizationType.applyRegularization(
```

Comment:
I don't think that is the correct way to calculate the updated weights with regularization. Regularization is part of the cost function we want to minimize and thus it influences our gradient. We have to incorporate the gradient of the regularization term into the overall gradient before calculating the new weights.

As far as I can tell, the L2 regularizer only need the old weight vector to calculate the regularizer gradient. The L1 regularizer using a proximal operator however operates on the new weight vector which it shrinks towards 0. Thus, in order to make the data treatment consistent, we could let the regularizer do the calculation of the updated weight vector. As input it gets the old weight vector, the gradient of the loss function, the stepsize and the regularization value. 

That way, it gives us enough flexibility to implement the L2 regularizer as well as the L1 regularizer.